creme/metrics/_confusion_matrix.pyx

```diff
@@ -0,0 +1,286 @@
+        self._init_classes = set(classes) if classes is not None else set()
+        self.sum_diag = 0.0
+        self.sum_row = defaultdict(float)
+        self.sum_col = defaultdict(float)
+        self.data = defaultdict(functools.partial(defaultdict, float))
+        self.n_samples = 0
+        self._class_counter = Counter()
+        self.last_y_true = None
+        self.last_y_pred = None
+
+    def __getitem__(self, key):
+        """Syntactic sugar for accessing the counts directly."""
+        return self.data[key]
+
+    def update(self, y_true, y_pred, sample_weight=1.):
+        self.n_samples += sample_weight
+
+        if sample_weight > 0:
+            self._class_counter.update([y_true])
+            # Keep track of last entry
+            self.last_y_true = y_true
+            self.last_y_pred = y_pred
+        else:
+            self._class_counter.subtract([y_true])
+
+        self.data[y_true][y_pred] += sample_weight
+
+        if y_true == y_pred:
+            self.sum_diag += sample_weight
+        self.sum_row[y_true] += sample_weight
+        self.sum_col[y_pred] += sample_weight
+
+        return self
+
+    def revert(self, y_true, y_pred, sample_weight=1.):
+        # Revert is equal to subtracting so we pass the negative sample_weight
+        self.update(y_true, y_pred, -sample_weight)
+        return self
+
+    @property
+    def classes(self):
+        return (
+            set(c for c, n in self.sum_row.items() if n) |
+            set(c for c, n in self.sum_col.items() if n)
+        )
+
+    @property
+    def n_classes(self):
+        return len(self.classes)
+
+    @property
+    def shape(self):
+        return self.n_classes, self.n_classes
+
+    def reset(self):
+        self.__init__(classes=self._init_classes)
+
+    def __repr__(self):
+
+        # The classes are sorted alphabetically for reproducibility reasons
+        classes = sorted(self.classes)
+
+        if not classes:
+            return ''
+
+        # Determine the required width of each column in the table
+        largest_label_len = max(len(str(c)) for c in classes)
+        largest_number_len = len(str(max(max(v for v in c.values()) for c in self.data.values())))
+        width = max(largest_label_len, largest_number_len) + 2
+
+        # Make a template to print out rows one by one
+        row_format = '{:>{width}}' * (len(classes) + 1)
+
+        # Write down the header
+        table = row_format.format('', *map(str, classes), width=width) + '\n'
+
+        # Write down the true labels row by row
+        table += '\n'.join((
+            row_format.format(
+                str(y_true),
+                *[f'{self.data[y_true][y_pred]:{self._fmt}}' for y_pred in classes],
+                width=width
+            )
+            for y_true in classes
+        ))
+
+        return table
+
+    @property
+    def majority_class(self):
+        try:
+            return self._class_counter.most_common(1)[0][0]
```

Comment:
We can also use the `sum_row` attribute, right? That way we don't need to create and update `_class_counter`.

Reply:
That was the original idea. I used the counter to avoid the loop to get the majority class. However, since we are using Cython this shouldn't be a big issue. Now we are using `sum_row`